content/courses/advanced-ada/parts/data_types/types.rst

```diff
@@ -6,13 +6,293 @@ Types
+can refer to a piece of data stored in memory |mdash| but it can also refer to
+a table or a form in a database. Also, even when we define the term *object* as
+data in memory, we can still classify programming languages as
+:wikipedia:`object-based <Object-based_language>` or
+:wikipedia:`object-oriented <Object-oriented_programming>` languages.
+
+.. admonition:: Important
+
+    In object-oriented programming, an object belongs to a *class* of objects.
+    In Ada, this kind of objects are called *tagged* objects. Note, however,
+    that we can have objects that don't belong to a class of objects: those are
+    called *untagged* objects.
+
+In the context of Ada programming, an object is an "entity that contains a
+value, and is either a constant or a variable" |mdash| according to the Ada
+Reference Manual. In other words, any constants or variables that we declare in
+Ada source code are objects. In addition, there are other examples of objects
+that don't originate from object declarations:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Data_Types.Types.Objects.Object_Examples
+
+    procedure Show_Objects is
+
+       type New_Integer is new
+         Integer;
+
+       type Integer_Array is
+         array (Positive range <>) of
+           Integer;
+
+       procedure Dummy (Obj : Integer)
+         is null;
+       --               ^^^
+       --              object
+
+       task type TT is
+          entry Start (Id : Integer);
+          --           ^^
+          --         object
+       end TT;
+
+       task body TT is
+       begin
+          accept Start (Id : Integer) do
+             null;
+          end Start;
+       end TT;
+
+       function Add_One (V : Integer)
+       --                ^
+       --    view of an object
+                         return Integer is
+       begin
+          return V + 1;
+          --     ^^^^^
+          --     object
+       end Add_One;
+
+       Arr : Integer_Array (1 .. 10);
+       --  ^^^^^^^^^^^^^^^
+       --  object
+
+       NI  : New_Integer;
+    begin
+       Arr (1 .. 3) := (others => 1);
+       --  ^^^^^^^^
+       --  object
+       --              ^^^^^^^^^^^^^
+       --                 object
+
+       NI := New_Integer (Arr (1));
+       --    ^^^^^^^^^^^^^^^^^^^^^
+       --      object
+
+       for I in Arr'Range loop
+       --  ^
+       --  object
+
+          Arr (I) := Add_One (Arr (I));
+          --                  ^^^^^^^
+          --                  object
+       end loop;
+    end Show_Objects;
+
+As we can see in this code example a formal parameter of a subprogram or an
+entry is also an object |mdash| in addition, so are
+:ref:`value conversions <Adv_Ada_Value_Conversion>`, the result returned by a
+function, the result of evaluating an :doc:`aggregate <./aggregates>`, loop
+parameters, :doc:`arrays <./arrays>`, or the slices or the components of an
+array.
+
```

Comment:
A little awkward here.
Instead of "or the slices or the components of an array." 
maybe "or the slices or the components of array objects."
But that should include record components too, so maybe
"or the slices of arrays objects, or the components of composite objects."

Reply:
Thanks for the suggestion! (I'm now using the last version.)